### DISCUSSION TITLE
Confused about testing.

### DISCUSSION THREAD

**Discussion by siph** [nushell/nupm#58] [Q&A]
In the past I was able to write tests in the same file as the script and run them with `use std testing run-tests ; run-tests`. I don't recall this being a sanctioned method, but it worked and was really convenient for me to include nushell tests into a nix-based ci (production warnings be damned). 

However, this method has now been deprecated and `nupm` is recommended instead.

```
Warning:   ⚠ deprecated_module
| the `std testing run-tests` command is deprecated and will be removed in Nushell 0.90

help: please use [`nushell/nupm`](https://github.com/nushell/nupm)

| Note
| Nupm is still a work in progress and is NOT production-ready, just as `run-tests` is.
| Please keep in mind that Nupm will change a lot in the near future and we can't recommend it for now.
| However, if you like bleeding edge software and want to give it a spin, we would welcome feedbacks and ideas :pray:
```

Based on the readme this workflow no longer seems possible. Instead I need to package the application for `nupm`.

I don't really mind this in many instances and I've been wanting to check out `nupm` for awhile, but it can make an already over-engineered [weather widget](https://github.com/siph/wt-fetch) even worse. Additionally, it requires a not-insignificant amount of overhead in both code and knowledge of `nupm` for the same functionality as before (still better than writing bash, though).

I skimmed the source code and see that I can export the tests and do something hacky with `scope commands` but this isn't ideal and as far as I can tell things like `#[before-all]` wouldn't work. I see the [nushell ci](https://github.com/nushell/nushell/blob/a093e668220a301b7fedc62f29d7fabecfae4105/.github/workflows/ci.yml#L128) is still using the old method, is this going to be changed or does nushell still have access to the `testing` module (if so, I'm curious why it was hidden)?

I'm just curious if there is a current or planned method of testing a single `.nu` file. Maybe [`nu_scripts`](https://github.com/nushell/nu_scripts) is a better place for something like this?

I'm hope that in the future we will enjoy some nice nix-nupm interoperability (I would be happy to be involved), but until then I would prefer to _not_ use `nupm` in most cases and just have some lower-level functionality that I can include into a nix build.

**Comment by amtoine** [comment]
the `std testing run-tests` command has only been hidden inside the standard library, to avoid rewriting things that work for now on the Nushell side.

i see two directions for this
- move the `std testing run-tests` to the `nu_scripts` as-is, to allow folks to still use the features it provides
- port some of the features of `std testing run-tests` to `nupm test`

**Reply by siph** [reply]
I personally would prefer the first option, especially if `nupm` may require a nightly build

**Reply by amtoine** [reply]
it's not really in my plans to continue supporting two versions of the same feature officially, so please feel free to add it to the `nu_scripts`, that would be reviewed and merged for sure :yum:
